js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext/util.ts

```diff
@@ -218,3 +221,54 @@ export function repoLocationToRepos(repositoryLocation: RepositoryLocation) {
     return {repository, repositoryLocation};
   });
 }
+
+export function mergeWorkspaceData(
+  workspaceLocationEntry: PartialWorkspaceLocationNodeFragment,
+  workspaceLocationAssetsEntry: WorkspaceLocationAssetsEntryFragment,
+) {
+  const result: Partial<WorkspaceLocationNodeFragment> = {};
+  const {locationOrLoadError, ...rest} = workspaceLocationEntry;
+  Object.assign(result, rest);
+  if (locationOrLoadError?.__typename === 'RepositoryLocation') {
+    const {repositories, ...rest} = locationOrLoadError;
+    result.locationOrLoadError = {
+      ...rest,
+      repositories: repositories
+        ? repositories.map((repo) => {
+            const assetsAndGroups = getAssetsAndGroupsByRepo(workspaceLocationAssetsEntry);
+            return {
+              ...repo,
+              ...(assetsAndGroups[repo.id] || {
+                assetNodes: [],
+                assetGroups: [],
+              }),
+            };
+          })
+        : [],
+    };
+  } else {
+    result.locationOrLoadError = locationOrLoadError;
+  }
+
+  return result as WorkspaceLocationNodeFragment;
+}
+
+function getAssetsAndGroupsByRepo(
+  workspaceLocationAssetsEntry: WorkspaceLocationAssetsEntryFragment,
+): Record<string, WorkspaceRepositoryAssetsFragment> {
+  if (workspaceLocationAssetsEntry.__typename !== 'WorkspaceLocationEntry') {
+    return {};
+  }
+
+  if (workspaceLocationAssetsEntry.locationOrLoadError?.__typename !== 'RepositoryLocation') {
+    return {};
+  }
+
+  return workspaceLocationAssetsEntry.locationOrLoadError.repositories.reduce(
+    (acc, repo) => {
+      acc[repo.id] = repo;
+      return acc;
+    },
+    {} as Record<string, WorkspaceRepositoryAssetsFragment>,
+  );
```

Comment:
You might be able to do this and avoid having to cast anything

```
const {repositories} = workspaceLocationAssetsEntry.locationOrLoadError;
return Object.fromEntries(repositories.map((repo) => [repo.id, repo]));
```

Over time I've found myself disliking `reduce` more... 😆